core/templates/pages/exploration-player-page/services/content-translation-manager.service.ts

```diff
@@ -0,0 +1,218 @@
+        card => this._displayTranslationsForCard(card, languageCode));
+    }
+
+    this.onStateCardContentUpdateEmitter.emit();
+  }
+
+  _swapContent(
+      writtenTranslations: WrittenTranslations,
+      languageCode: string,
+      content: SubtitledHtml|SubtitledUnicode
+  ): void {
+    const writtenTranslation = writtenTranslations.translationsMapping[
+      content.contentId][languageCode];
+
+    if (!this._isValidTranslation(writtenTranslation)) {
+      return;
+    }
+
+    let valueName;
+    if (content instanceof SubtitledHtml) {
+      valueName = 'html';
+    } else if (content instanceof SubtitledUnicode) {
+      valueName = 'unicode';
+    }
+
+    content[valueName] = writtenTranslation.translation;
+  }
+
+  _displayTranslationsForCard(card: StateCard, languageCode: string) : void {
+    const writtenTranslations = card.writtenTranslations;
+
+    const contentTranslation = writtenTranslations.translationsMapping[
+      card.contentId][languageCode];
+    if (this._isValidTranslation(contentTranslation)) {
+      card.contentHtml = contentTranslation.translation;
+    }
+
+    card.getHints().forEach(hint => this._swapContent(
+      writtenTranslations, languageCode, hint.hintContent));
+
+    const solution = card.getSolution();
+    if (solution !== null) {
+      this._swapContent(
+        writtenTranslations, languageCode, solution.explanation);
+    }
+
+    if (card.getInteraction().defaultOutcome) {
+      this._swapContent(
+        writtenTranslations,
+        languageCode,
+        card.getInteraction().defaultOutcome.feedback);
+    }
+
+    const answerGroups = card.getInteraction().answerGroups;
+    answerGroups.forEach(answerGroup => {
+      this._swapContent(
+        writtenTranslations, languageCode, answerGroup.outcome.feedback);
+    });
+
+    if (card.getInteraction().id) {
+      this._swapContentInCustomizationArgs(
+        card, languageCode);
+    }
+  }
+
+  _swapContentInCustomizationArgs(card: StateCard, languageCode: string): void {
+    const interactionId = card.getInteractionId();
+    const caValues = card.getInteraction().customizationArgs;
+    const writtenTranslations = card.writtenTranslations;
+
+    let traverseSchemaAndSwapTranslatableContent = (
+        value: Object | Object[],
+        schema: Schema
+    ): void => {
+      const schemaIsSubtitledHtml = (
+        schema.type === SchemaConstants.SCHEMA_TYPE_CUSTOM &&
+        schema.obj_type === SchemaConstants.SCHEMA_OBJ_TYPE_SUBTITLED_HTML);
+      const schemaIsSubtitledUnicode = (
+        schema.type === SchemaConstants.SCHEMA_TYPE_CUSTOM &&
+        schema.obj_type === SchemaConstants.SCHEMA_OBJ_TYPE_SUBTITLED_UNICODE
+      );
+
+      if (schemaIsSubtitledHtml || schemaIsSubtitledUnicode) {
+        this._swapContent(
+          writtenTranslations, languageCode,
+          <SubtitledHtml|SubtitledUnicode>value);
+      } else if (schema.type === SchemaConstants.SCHEMA_KEY_LIST) {
+        for (
```

Comment:
Would using forEach be better here, since you don't actually really make use of `i`? (Not sure.)

Reply:
Done!